index.js

```diff
@@ -30,56 +47,67 @@ module.exports = options => {
 			return;
 		}
 
-		const finish = (error, size) => {
+		const finish = (error, sizes) => {
 			if (error) {
 				callback(new PluginError('gulp-size', error));
 				return;
 			}
 
-			totalSize += size;
+			for (const [key, size] of sizes) {
+				totalSize.set(key, size + (totalSize.get(key) || 0));
+			}
 
-			if (options.showFiles === true && size > 0) {
-				log(chalk.blue(file.relative), size);
+			if (options.showFiles === true && hasSize(sizes)) {
+				log(chalk.blue(file.relative), sizes);
 			}
 
 			fileCount++;
 			callback(null, file);
 		};
 
+		const selectedSizes = new Map();
 		if (file.isStream()) {
+			if (options.uncompressed) {
+				selectedSizes.set('', promisify(file.contents.pipe(new StreamCounter()), 'bytes', 'finish'));
+			}
+
 			if (options.gzip) {
-				file.contents.pipe(gzipSize.stream())
-					.on('error', finish)
-					.on('end', function () {
-						finish(null, this.gzipSize);
-					});
-			} else {
-				file.contents.pipe(new StreamCounter())
-					.on('error', finish)
-					.on('finish', function () {
-						finish(null, this.bytes);
-					});
+				selectedSizes.set(' (gzipped)', promisify(file.contents.pipe(gzipSize.stream()), 'gzipSize'));
 			}
 
-			return;
+			if (options.brotli) {
+				selectedSizes.set(' (brotli)', promisify(file.contents.pipe(brotliSize.stream()), 'brotliSize'));
+			}
 		}
 
-		if (options.gzip) {
-			(async () => {
-				try {
-					finish(null, await gzipSize(file.contents));
-				} catch (error) {
-					finish(error);
-				}
-			})();
-		} else {
-			finish(null, file.contents.length);
+		if (file.isBuffer()) {
+			if (options.uncompressed) {
+				selectedSizes.set('', file.contents.length);
+			}
+
+			if (options.gzip) {
+				selectedSizes.set(' (gzipped)', gzipSize(file.contents));
+			}
+
+			if (options.brotli) {
+				selectedSizes.set(' (brotli)', brotliSize.default(file.contents));
```

Comment:
I don't think the output ` (brotli)` should be the key. That's too much coupling. Use proper keys like `'brotli'` and just translate it to the actual title when logging using a `Map`:

```js
new Map([
	['brotli', ' (brotli)'],
	…
])
```

Reply:
Agreed.